interactions/gateway.py

```diff
@@ -1,178 +1,256 @@
+        """
+        super().__init__()
+        self.websocket = websocket
+        self.interval = interval
+        self.event = Event()
 
-    CONNECT TO WEBSOCKET SERVER
-    |- OPCODE 10/HELLO
-       |- IDENTIFY
-          |- RESUME
-             |- HEARTBEAT
-             |- OPCODE 11 / HEARTBEAT_ACK
+    async def run(
+        self,
+        interval: float,
+        websocket: Any
+    ) -> None:
+        """
+        Automatically send heartbeats to the gateway.
+
+        .. note::
+
+            Inherits the arguments from the class constructor.
+
+        :return: None
+        """
+        while True:
+            sleep(interval)
+            await self.websocket.heartbeat()
+
+    def stop(self) -> None:
+        """
+        Stop sending heartbeats to the gateway.
+
+        :return: None
+        """
+        return self.event.set()
+
+class WebSocket:
+    """
+    Object representing logic for connecting to the gateway.
 
-    :ivar path:
     :ivar websocket:
     :ivar data:
-    :ivar token:
-    :ivar session_id:
     """
     __slots__ = (
-        "path",
-        "websocket",
-        "data",
-        "token",
-        "session_id"
+        "session",
+        "session_id",
+        "heart",
+        "sequence",
+        "interval",
+        "intents",
+        "closed"
     )
-    path: str
-    websocket: Any
-    data: Optional[dict]
-    token: str
+    session: Any
     session_id: Optional[int]
+    heart: Optional[Heartbeat]
+    sequence: Optional[int]
+    interval: Optional[int]
+    intents: Optional[int]
+    closed: bool
+
+    def __init__(self) -> None:
+        """
+        Instantiates the WebSocket class.
 
-    async def __init__(
+        :return: None
+        """
+        self.closed = False
+
+    async def connect(
         self,
-        token: str
+        intents: Optional[int]=513,
+        session_id: Optional[int]=None
     ) -> None:
         """
-        Instantiates the WebSocket class.
+        Validates a connection to the WebSocket server.
         
-        :param token: The token to use to connect to the gateway.
-        :type token: str
+        :param intents: The intents to pass for the application connection.
+        :type intents: typing.Optional[int]
         :return: None
         """
-        self.path = (Route.GATEWAY, "&encoding=json")
-        self.websocket = None
-        self.data = None
-        self.token = token
-        self.session_id = None
+        self.intents = intents
+        self.session_id = session_id
+        path: str = Route.GATEWAY.value + "&encoding=json"
+        async with websockets.connect(path) as self.session:
+            stream = await self.recv()
+            fields: dict = {
+                "op": stream.get("op"),
+                "data": stream.get("d"),
+                "seq": stream.get("s")
+            }
+            self.sequence: int = fields["seq"] if fields["seq"] is not None else "null"
+
+            while not self.closed:
```

Comment:
This logic is broken. You recieve one event, and then loop over this event over and over. Move this while loop to wrap line 130